### PR TITLE
API Call, Bloc, Testing and new UI Widgets

### PR DESCRIPTION
**Pull request by marcosvonstremayr**
In this module I incorporated new Widgets such as a Drawer and a GridView among others. All the cards information is fetched from the API of Hearthstone and the state is managed using the BLoC Pattern.

https://user-images.githubusercontent.com/109096201/187760737-021bd4b4-a8c8-4d08-b0c5-b31668384d01.mp4

image

### CODE REVIEW

File: lib/main.dart [hunk 1]
```diff
@@ -1,21 +1,63 @@
...
+
+  @override
+  void initState() {
+    super.initState();
+    _cardRequestParams = CardsRequestParams();
+    _apiService = CardsApiService();
+    _cardRepository = CardRepositoryImpl(_apiService);
+    _getCardsUsecase = GetCardsUsecase(_cardRepository);
+    _cardEvent = CardEvent([], Status.initial);
+    _bloc = CardsBlocImpl(_getCardsUsecase, _cardRequestParams, _cardEvent);
```
Comment (FranciscoGaraventa) [inline lib/main.dart]: trailing comma

File: lib/src/data/datasource/remote/cards_api_service.dart [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+  Future<dynamic> callApi(params) async {
+    Uri uri;
+    if (params.endpoint != null) {
+      uri = Uri.parse(params.apiBaseUrl + params.endpoint);
+    } else {
+      uri = Uri.parse(params.apiBaseUrl);
+    }
+    try {
+      http.Response response = await _client.get(uri, headers: params.apiHeader);
```
Comment (FranciscoGaraventa) [inline lib/src/data/datasource/remote/cards_api_service.dart]: Instead to http.Response you can use var here

File: lib/src/data/datasource/remote/cards_api_service.dart [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+      if (response.statusCode == 200) {
```
Comment (FranciscoGaraventa) [inline lib/src/data/datasource/remote/cards_api_service.dart]: What happens if the answer is different from 200?

Comment (FranciscoGaraventa) [reply lib/src/data/datasource/remote/cards_api_service.dart]: Instead to 200, you can use HttpStatus.ok

File: lib/src/domain/entity/card_event.dart [hunk 4]
```diff
@@ -0,0 +1,14 @@
+import '../../data/model/card_model.dart';
+
+class CardEvent {
+  List<CardModel> cards;
+  Status status;
+
+  CardEvent(this.cards, this.status);
```
Comment (FranciscoGaraventa) [inline lib/src/domain/entity/card_event.dart]: trailing comma

File: lib/src/presentation/bloc/cards_bloc/cards_bloc_impl.dart [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+import '../../../domain/entity/card_event.dart';
+import '../../../core/usecases/usecase.dart';
+import 'cards_bloc.dart';
+
+class CardsBlocImpl extends CardsBloc {
+  final Usecase _getCardsUsecase;
+  CardsRequestParams _cardsRequestParams;
+  CardEvent _cardEvent;
+
+  CardsBlocImpl(this._getCardsUsecase, this._cardsRequestParams, this._cardEvent);
```
Comment (FranciscoGaraventa) [inline lib/src/presentation/bloc/cards_bloc/cards_bloc_impl.dart]: trailing comma

File: lib/src/presentation/bloc/cards_bloc/cards_bloc_impl.dart [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+  void dispose() {
+    _cardsStreamController.close();
+  }
+
+  @override
+  Future<void> getFilteredCards(endpoint) async {
+    _cardEvent.status = Status.loading;
+    _cardsStreamController.sink.add(_cardEvent);
+    _cardEvent.cards.clear();
+    if (endpoint != null) {
```
Comment (FranciscoGaraventa) [inline lib/src/presentation/bloc/cards_bloc/cards_bloc_impl.dart]: What it the scenario that allows an endpoint be null?

File: lib/src/presentation/view/homepage.dart [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+
+import '../../core/util/constants.dart';
+import '../../domain/entity/card_event.dart';
+import '../bloc/cards_bloc/cards_bloc.dart';
+import '../widget/grid_text.dart';
+import '../widget/show_img.dart';
+import 'card_detail.dart';
+
+class Homepage extends StatelessWidget {
+  const Homepage({Key? key, required this.bloc}) : super(key: key);
```
Comment (FranciscoGaraventa) [inline lib/src/presentation/view/homepage.dart]: trailing comma

File: lib/src/presentation/view/homepage.dart [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+                  return const CircularProgressIndicator();
+                } else {
+                  return GridView.builder(
+                    gridDelegate:
+                        const SliverGridDelegateWithFixedCrossAxisCount(
+                      crossAxisCount: Constants.numberOfGridColumns,
+                      childAspectRatio: Constants.gridChildAspectRatio,
+                    ),
+                    itemCount: snapshot.data!.cards.length,
+                    itemBuilder: (BuildContext context, int index) {
```
Comment (FranciscoGaraventa) [inline lib/src/presentation/view/homepage.dart]: trailing comma

File: lib/src/presentation/widget/show_img.dart [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+      img,
+      height: height,
+    );
+  } else {
+    return Image.asset(
+      Constants.backCardImage,
+      height: height,
+    );
+  }
+}
```
Comment (FranciscoGaraventa) [inline lib/src/presentation/widget/show_img.dart]: Instead to do this widget, you can use CachedNetworkImage

File: test/unit_tests/api_service/cards_api_service_test.dart [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+        client.get(
+          Uri.parse(
+            params.apiBaseUrl + params.endpoint!,
+          ),
+          headers: params.apiHeader,
+        ),
+      ).thenAnswer((_) async {
+        return http.Response(
+          jsonEncode(cardsTestsJsonFilteredCards),
+          200,
```
Comment (FranciscoGaraventa) [inline test/unit_tests/api_service/cards_api_service_test.dart]: Here you can user HttpStatus.ok

File: test/unit_tests/api_service/cards_api_service_test.dart [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+    test("Api Service Exception test", () async {
+      when(
+        client.get(
+          Uri.parse(params.apiBaseUrl),
+          headers: params.apiHeader,
+        ),
+      ).thenAnswer((_) async {
+        return http.Response(
+          jsonEncode(cardsTestEmptyJson),
+          404,
```
Comment (FranciscoGaraventa) [inline test/unit_tests/api_service/cards_api_service_test.dart]: Here you can user HttpStatus.notFound

File: test/unit_tests/request_params_unit_test.dart [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+  setUp(() {
+    requestParams = CardsRequestParams();
+  });
+
+  group("Card Request Params Unit Testing", () {
+    test("Class instantiation", () {
+      expect(requestParams, isA<CardsRequestParams>());
+    });
+  });
+}
```
Comment (FranciscoGaraventa) [inline test/unit_tests/request_params_unit_test.dart]: Add new line at EOF

File: test/unit_tests/usecase/get_cards_usecase_test.dart [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+      when(cardRepository.fetchFilteredCards(params)).thenAnswer((_) async{
+        return cardModelList;
+      });
+      expectLater(await getCardsUsecase(params: params), isA<List<CardModel>>());
+    });
+  });
+
+
+
+}
```
Comment (FranciscoGaraventa) [inline test/unit_tests/usecase/get_cards_usecase_test.dart]: Remove this extras lines and add new line at EOF

File: lib/main.dart (hunk 1 above)
Comment (marcosvonstremayr) [reply lib/main.dart]: Done

File: lib/src/data/datasource/remote/cards_api_service.dart (hunk 2 above)
Comment (marcosvonstremayr) [reply lib/src/data/datasource/remote/cards_api_service.dart]: Done

File: lib/src/data/datasource/remote/cards_api_service.dart (hunk 3 above)
Comment (marcosvonstremayr) [reply lib/src/data/datasource/remote/cards_api_service.dart]: I modified the code to throw an exception when the response.statusCode != HttpStatus.ok. So the catch block will be executed, is that ok?

File: lib/src/presentation/bloc/cards_bloc/cards_bloc_impl.dart (hunk 6 above)
Comment (marcosvonstremayr) [reply lib/src/presentation/bloc/cards_bloc/cards_bloc_impl.dart]: You are right, there is no scenario

File: lib/src/presentation/view/homepage.dart (hunk 7 above)
Comment (marcosvonstremayr) [reply lib/src/presentation/view/homepage.dart]: Done

File: lib/src/presentation/view/homepage.dart (hunk 8 above)
Comment (marcosvonstremayr) [reply lib/src/presentation/view/homepage.dart]: Done

File: lib/src/presentation/widget/show_img.dart (hunk 9 above)
Comment (marcosvonstremayr) [reply lib/src/presentation/widget/show_img.dart]: Done

File: test/unit_tests/api_service/cards_api_service_test.dart (hunk 10 above)
Comment (marcosvonstremayr) [reply test/unit_tests/api_service/cards_api_service_test.dart]: Done

File: test/unit_tests/api_service/cards_api_service_test.dart (hunk 11 above)
Comment (marcosvonstremayr) [reply test/unit_tests/api_service/cards_api_service_test.dart]: Done

File: test/unit_tests/request_params_unit_test.dart (hunk 12 above)
Comment (marcosvonstremayr) [reply test/unit_tests/request_params_unit_test.dart]: Done

File: test/unit_tests/usecase/get_cards_usecase_test.dart (hunk 13 above)
Comment (marcosvonstremayr) [reply test/unit_tests/usecase/get_cards_usecase_test.dart]: Done

File: lib/src/core/util/constants.dart [hunk 14]
```diff
@@ -3,65 +3,106 @@ import 'package:flutter/material.dart';
 abstract class Constants {
```
Comment (FranciscoGaraventa) [inline lib/src/core/util/constants.dart:3]: This constants file could be divided  in different files, one relative to dimens (Like fontSize, height, Colors), the second one relative to String constants, another relative to api service calls (apiBaseUrl, api endpoints ..), and other only to assets.

**Review by FranciscoGaraventa — APPROVED** [review]
Approve with minor comment

File: lib/src/core/util/constants.dart (hunk 14 above)
Comment (marcosvonstremayr) [reply lib/src/core/util/constants.dart:3]: Done

**Review by LeonardoMolleker — APPROVED** [review]

**Review by MarcosCGlobant — APPROVED** [review]

**Pull request merged**